### DISCUSSION TITLE
Use roles without relying on them being set up in OpenID Connect

### DISCUSSION THREAD

**Discussion by carwei** [lowdefy/lowdefy#1412] [Questions and Help] [answered]
I want to enable login and roles for an app, but I don’t have control over the claims I’m OpenID Connect. Is there a work around for this? 

I would like to manage the roles in my own database. Having tables with all valid users, all possible roles and a join table with who has which role.

The flow would be something like this:
1. User visits app, clicks “log in”
2. Redirected to SSO page, then back to callback url for the app
3. The backend of the app verifies that the “SSO jwt” is valid (not tampered with) and the email field matches a user’s email in the database
4. The backend gets the user’s roles from the database and creates its own “app jwt” containing both user data from the SSO jwt and the roles from the database, signed with a jwt secret (not necessarily the same secret used be OpenID Connect). This would replace the original token.
5. Lowdefy plays along like the claims came straight from OpenID Connect…? *crossing my fingers*

Would this work? Would you suggest any alternative?
Thanks in advance!

**Comment by SamTolmay** [comment, marked as answer]
Hi @carwei 

In V4 use can use auth adapter plugins to store user data and sessions in your own DB. These are based on https://next-auth.js.org

You can have a look at this example repo where we implement user invites and roles in MongoDB: https://github.com/lowdefy/starter-auth0-mongo

You should be able to write a similar adapter function that interfaces with your DB. You can also look at the default next-auth adapters to use as a starting point: https://github.com/nextauthjs/next-auth/tree/main/packages

**Reply by carwei** [reply]
Ok, really looking forward to V4!

**Reply by carwei** [reply]
Couldn't find Postgres in the list of official adapters, but this looks promising: https://www.jakobmaier.at/posts/next-auth-postgres-adapter/
